subsys/bluetooth/host/settings.c

```diff
@@ -174,6 +174,22 @@ static int set(const char *name, size_t len_rd, settings_read_cb read_cb,
 	}
 #endif
 
+#if defined(CONFIG_BT_DEVICE_APPEARANCE_DYNAMIC)
+	if (!strncmp(name, "appearance", len)) {
+		if (len != sizeof(bt_dev.appearance)) {
+			BT_ERR("Ignoring settings entry 'bt/appearance'. Wrong length.");
```

Comment:
1) We should definitely use a `#define or a static const variable to store `bt/appearance`. Relying on a string literal being the same in two different files sound like a recipe for disaster. 
2) You do a strncmp against `appearance`, but save it as `bt/appearance`. Is that intentional?

Reply:
For 1. I see that's also the case for the others, so it is OK for this PR, but if we don't fix that in this PR, I strongly suggest that we create a GH issue for it.

Reply:
> 1. We should definitely use a #define or a static const variable to store bt/appearance`. Relying on a string literal being the same in two different files sound like a recipe for disaster.

I agree. I wanted to make define a symbol, but the fact that the key name is stripped in function makes it not straight forward. Should we have two symbols? `BT_SETTINGS_KEY_PREFIX` and `BT_SETTINGS_KEY_APPEARANCE`? It gets very verbose. Should `BT_SETTINGS_KEY_PREFIX` have a `/` at the end? I'm not sure this makes the situation any less error prone.

> 2. You do a strncmp against appearance, but save it as bt/appearance. Is that intentional?

Yes. The `name` argument to this handler is stripped of the subtree prefix "bt/".

Reply:
> Should we have two symbols? BT_SETTINGS_KEY_PREFIX and BT_SETTINGS_KEY_APPEARANCE? 

If we strip the `bt/` prefix, then I guess we need 2 symbols. Why do we strip that btw? If it makes it easier to compare, then it would be easier to just not remove that symbol. 

Alternatively, we could have a `bt_settings_save_one` that automatically adds the `bt/` prefix so that e.g.
```
err = settings_save_one("bt/name", bt_dev.name, len);
```
becomes
```
err = bt_settings_save_one("name", bt_dev.name, len);
```

and then we can just have one symbol to compare against. The reading symbol (e.g. `"name"`), should be the save as the storing symbol (which currently is `"bt/name"`), so we can either omit the `bt/` prefix both places, or explicitly add it both places, instead of this weird hybrid we have now.

Reply:
For context, the host pretty much follows the Settings API as intended. See the following example: https://docs.zephyrproject.org/latest/reference/settings/index.html#example-persist-runtime-state. The same problem is present there.

A `bt_settings_save_one` is not a bad idea. It could be implemented as a macro that adds "bt/" at compile time. But again, it's easy to forget to use `bt_settings_save_one` instead of `settings_save_one`. (We could define BT_SETTINGS_KEY_APPEARANCE as a token sequence that does not fit into `settings_save_one`, but I feel it is too much of a hack.)

Alternatively, we could go the other way and remove the leading "bt/" using `settings_name_next`.

Reply:
> For context, the host pretty much follows the Settings API as intended. See the following example: https://docs.zephyrproject.org/latest/reference/settings/index.html#example-persist-runtime-state. The same problem is present there.

Indeed, that seems to be a generic issue then. Given the example in that, it would almost seem that `settings_save_one` should be refactored from
```
int settings_save_one(const char *name, const void *value, size_t val_len)
```
into 
```
int settings_save_one(const char *name, const char *value_name, const void *value, size_t val_len)
```

so that
```
settings_save_one("foo/bar", &foo_val, sizeof(foo_val));
```
becomes 
```
settings_save_one("foo", "bar", &foo_val, sizeof(foo_val));
```
and thus the `struct settings_handler` uses the same `.name` as the function. 

> But again, it's easy to forget to use bt_settings_save_one instead of settings_save_one.

I agree to some extend. You can also argue that it is possible to use the generic function instead of the specific function, but that's almost the same as arguing that it is easy to use `bt_hci_cmd_send_sync` instead of e.g. e.g. the `bt_le_*` functions.

Reply:
I like your proposal to refactor. I would just change the parameter names a bit:
```c
int settings_save_one(const char *subtree_name, const char *key_name, const void *value, size_t value_len);
```

But this is a user-facing API change, so we definitely should do that in a separate issue. For now, I'll add a #define for "foo" and "bar". I'll push it up right away.

Reply:
I'll go ahead and open a GH issue to enhance this :)

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/42048